README.md
# Parse-CPAN-Meta has merged into CPAN-Meta

In order to simplify development of these tightly coupled modules,
Parse-CPAN-Meta has merged into CPAN-Meta, available as
https://github.com/Perl-Toolchain-Gang/CPAN-Meta

This repository is being kept alive for the commit and issue histories.  No
further development will occur in this repository.  All new issues should
be filed with CPAN-Meta.


Changes
Changes for Perl programming language extension Parse-CPAN-Meta

{{$NEXT}}

1.4422    2016-07-02 21:02:57-04:00 America/New_York

    [Fixed]

    - Fixed use of Encode in load_json_string.

1.4421    2016-06-28 16:59:09-04:00 America/New_York

    - No changes from 1.4420

1.4420    2016-06-23 07:07:05-04:00 America/New_York (TRIAL RELEASE)

    [TESTS]

    - Adds tests cases for wide characters in META files.

    [PREREQS]

    - Bumps JSON::PP prerequisite to 2.27300 to work around a bug before
      perl 5.8.7.  Includes a test to confirm correct behavior.

1.4419    2016-04-30 15:32:12+00:00 UTC (TRIAL RELEASE)

    - Re-encode strings before decode_json since that expects bytes

1.4418    2016-04-24 13:15:53+00:00 UTC (TRIAL RELEASE)

    - Add CPAN_META_JSON_BACKEND to allow requesting non-JSON.pm backends
    - Add CPAN_META_JSON_DECODER to allow Mojo::JSON/JSON::Tiny to be used
      just for decoding.

1.4417    2015-06-09 16:19:41-06:00 America/Denver

    - No changes from 1.4416

1.4416    2015-05-19 11:11:47-04:00 America/New_York (TRIAL RELEASE)

    [FIXED]

    - Minimum Perl was inadvertently set to v5.10.0.  Now back to v5.8.1.

1.4415    2015-04-28 11:29:52-04:00 America/New_York (TRIAL RELEASE)

    [TESTS]

    - Outputs the version of backends used

    [META]

    - Updated repo metadata and boilerplate files

    - Pointed issue tracker to the Perl-Toolchain-Gang Github repo

1.4414    2014-03-04 15:22:00-05:00 America/New_York

    [ADDED]

    - Added load_string, which uses heuristics to guess YAML/JSON content

    [CHANGED]

    - load_file will now attempt to detect YAML/JSON content if the suffix
      is not .json, .yml or .yaml

1.4413    2014-02-17 20:04:23-05:00 America/New_York

    [FIXED]

    - UTF-8 decoding is done differently to avoid requiring
      a newer version of Encode (Graham Knop)

1.4412    2014-02-17 07:14:43-05:00 America/New_York

    [FIXED]

    - Added prerequisite on Encode 2.11 (Paul Howarth)

1.4411    2014-02-16 23:06:00-05:00 America/New_York

    [FIXED]

    - Invalid UTF-8 encoding in YAML files are now replaced with "PERLQQ"
      quoting from the Encode module and without warnings.

    - Removed legacy test modifications for $ENV{PERL_CORE}

1.4410    2014-02-16 00:47:48-05:00 America/New_York

    [INCOMPATIBLE CHANGES]

    - Dropped support for Perl's before v5.8.1 to ensure proper
      UTF-8 and IO layer support

    [CHANGED]

    - Updated CPAN::Meta::YAML prereq to 0.011 due to deprecation
      of 'errstr' and consistent use of exceptions

1.4409    2013-09-23 13:17:18 America/New_York

    [FIXED]

    - updated Makefile.PL logic to support PERL_NO_HIGHLANDER

1.4408    2013-09-23 11:06:35 America/New_York

    [FIXED]

    - Installation on Perls < 5.12 will uninstall older versions installed
      due to being bundled with ExtUtils::MakeMaker

1.4407    2013-09-08 10:16:29 America/New_York

    - Fixed incorrect resource metadata

1.4406    2013-09-08 09:59:28 America/New_York

    - Fixed improperly encoded META.yml

1.4405    2013-07-17 21:43:34 America/New_York

    - Fixed incorrect "return ... or die ..." constructs

    - Converted distribution to Dist::Zilla management

    - Noted new repository location in the Github Perl-Toolchain-Gang
      organization

1.4404 2012-04-05

    - Protected tests from user PERL_YAML/JSON_BACKEND

1.4403 2012-04-01

    - Bumped prereqs: JSON::PP to 2.27200 and CPAN::Meta::YAML to 0.008

1.4402 2012-02-07

    - Minor maintenance: standardized newlines to Unix style

1.4401 2011-02-04

    - Removed Module::Load::Conditional dependency

1.4400 2011-02-04

    - Added 'json_backend' and 'yaml_backend' methods to provide
    the names of modules used for deserialization

1.4200 2011-01-24

    - No changes from 1.41_04

1.41_04 2011-01-03

    - Support PERL_JSON_BACKEND environment (defaulting to JSON::PP)

    - Support PERL_YAML_BACKEND environment (defaulting to CPAN::Meta:YAML)

    - Update Makefile.PL to install to sitelib on Perl 5.12+

1.41_03 2010-12-17

    - Throw exception when CPAN::Meta::YAML has a parse error
      (RT#47608)

1.41_02 2010-12-17

    - Convert to using CPAN::Meta::YAML instead of YAML::Tiny

1.41_01 2010-12-10

    - add support for JSON metafiles and load_* methods (RJBS)

    - Move binary file unpacking to test file from Makefile.PL so tests
      will pass in the Perl core using a generated Makefile.PL

    - Move bundled uupacktool.pl to t/bin/

1.40 2009-07-25

    - Add core perl 5.10.1's uupacktool.pl

    - Repackage t/data/utf_16_le_bom.yml as ASCII for https://rt.cpan.org/Ticket/Display.html?id=47844

1.39 2009-05-21

    - Even though utf8 starts at 5.7+ there's no is_utf till
      5.8.1 so skip in the tests if needed (ADAMK)

1.38 2009-05-16

    - Synchorised version numbers with YAML::Tiny

    - Code shrinkages to the bloaty Unicode/BOM code

    - Merging a ton more tests from YAML::Tiny

    - Added back some pointless options to yaml_ok that keep our
      codebase closer to YAML::Tiny (which should help with syncing)

    - Changed exceptions to be closer to YAML::Tiny

0.05 2009-03-11

    - Upgrade previous release to a stable release.

0.04_01 2009-03-03

    - Merge Unicode / BOM fixes from YAML::Tiny

0.04 2009-01-07

    - Matching changes in YAML::Tiny 1.36

    - Fixing missing feature reported by H.Merijn Brand

    - Changes to make Padre-CPAN-Meta core-compatible (SMUELLER)++

0.03 2008-03-20

    - Adding YAML.pm compatibility in scalar context

0.02 2008-01-10

    - Purging some references to YAML::Tiny that I missed

0.01 2008-01-08

    - Cloned from YAML::Tiny 1.21

    - Removed all write functionality

    - Reduced interface to Load and LoadFile

    - Removed object-orientation

    - Removed global $errstr, all errors are exceptions

    - Applied prototypes to all functions


t/lib/Test.pm
package t::lib::Test;

use strict;
use Test::More ();
use Parse::CPAN::Meta;

use vars qw{@ISA @EXPORT};
BEGIN {
	require Exporter;
	@ISA    = qw{ Exporter };
	@EXPORT = qw{ tests  yaml_ok  slurp  load_ok };
}

# 22 tests per call to yaml_ok
# 4  tests per call to load_ok
sub tests {
	return ( tests => count(@_) );
}

sub count {
	my $yaml_ok = shift || 0;
	my $load_ok = shift || 0;
	my $single  = shift || 0;
	my $count   = $yaml_ok * 3 + $load_ok * 4 + $single;
	return $count;
}

sub yaml_ok {
	my $string  = shift;
	my $array   = shift;
	my $name    = shift || 'unnamed';

	# Does the string parse to the structure
	my $yaml_copy = $string;
	my @yaml      = eval { Parse::CPAN::Meta::Load( $yaml_copy ); };
	Test::More::is( $@, '', "$name: Parse::CPAN::Meta parses without error" );
	Test::More::is( $yaml_copy, $string, "$name: Parse::CPAN::Meta does not modify the input string" );
	SKIP: {
		Test::More::skip( "Shortcutting after failure", 1 ) if $@;
		Test::More::is_deeply( \@yaml, $array, "$name: Parse::CPAN::Meta parses correctly" );
	}

	# Return true as a convenience
	return 1;
}

sub slurp {
	my $file = shift;
	local $/ = undef;
	open( FILE, " $file" ) or die "open($file) failed: $!";
	my $source = <FILE>;
	close( FILE ) or die "close($file) failed: $!";
	$source;
}

sub load_ok {
	my $name = shift;
	my $file = shift;
	my $size = shift;
	Test::More::ok( -f $file, "Found $name" );
	Test::More::ok( -r $file, "Can read $name" );
	my $content = slurp( $file );
	Test::More::ok( (defined $content and ! ref $content), "Loaded $name" );
	Test::More::ok( ($size < length $content), "Content of $name larger than $size bytes" );
	return $content;
}

1;


t/lib/Parse/CPAN/Meta/Test.pm
package Parse::CPAN::Meta::Test;

use strict;
use Test::More ();
use Parse::CPAN::Meta;
use File::Spec;

use vars qw{@ISA @EXPORT};
BEGIN {
	require Exporter;
	@ISA    = qw{ Exporter };
	@EXPORT = qw{ tests  yaml_ok  slurp  load_ok  test_data_directory };
}

sub test_data_directory {
	return(
		$ENV{PERL_CORE}
		? File::Spec->catdir(qw(lib Parse CPAN Meta t data))
		: File::Spec->catdir(qw(t data))
	);
}

# 22 tests per call to yaml_ok
# 4  tests per call to load_ok
sub tests {
	return ( tests => count(@_) );
}

sub count {
	my $yaml_ok = shift || 0;
	my $load_ok = shift || 0;
	my $single  = shift || 0;
	my $count   = $yaml_ok * 3 + $load_ok * 4 + $single;
	return $count;
}

sub yaml_ok {
	my $string  = shift;
	my $array   = shift;
	my $name    = shift || 'unnamed';

	# Does the string parse to the structure
	my $yaml_copy = $string;
	my @yaml      = eval { Parse::CPAN::Meta::Load( $yaml_copy ); };
	Test::More::is( $@, '', "$name: Parse::CPAN::Meta parses without error" );
	Test::More::is( $yaml_copy, $string, "$name: Parse::CPAN::Meta does not modify the input string" );
	SKIP: {
		Test::More::skip( "Shortcutting after failure", 1 ) if $@;
		Test::More::is_deeply( \@yaml, $array, "$name: Parse::CPAN::Meta parses correctly" );
	}

	# Return true as a convenience
	return 1;
}

sub slurp {
	my $file = shift;
	local $/ = undef;
	open( FILE, " $file" ) or die "open($file) failed: $!";
	my $source = <FILE>;
	close( FILE ) or die "close($file) failed: $!";
	$source;
}

sub load_ok {
	my $name = shift;
	my $file = shift;
	my $size = shift;
	Test::More::ok( -f $file, "Found $name" );
	Test::More::ok( -r $file, "Can read $name" );
	my $content = slurp( $file );
	Test::More::ok( (defined $content and ! ref $content), "Loaded $name" );
	Test::More::ok( ($size < length $content), "Content of $name larger than $size bytes" );
	return $content;
}

1;


diff_to_blead.pl
#!/usr/bin/env perl
use strict;
use warnings;
use Getopt::Long;

my $diff = 'diff';

my $lib  = 'lib/Parse/CPAN';
my $tlib = 't/lib/Parse/CPAN/Meta';
# local => blead
my %files = (
  "$lib/Meta.pm" => "$lib/Meta.pm",

  "$tlib/Test.pm" => "$tlib/Test.pm",
  "Changes" => "$lib/Meta/Changes",

  (
    map {my $f = "t/$_.t"; ($f, "$lib/Meta/$f")}
    qw( 01_compile 02_basic 03_regression 04_scalar
        11_meta_yml 12_plagger 13_perl_smith 14_yaml_org
        15_multibyte 16_nullrefs 17_toolbar )
  ),
  (
    map {my $f = "t/data/$_"; ($f, "$lib/Meta/$f")}
    qw( HTML-WebDAO.yml multibyte.yml sample.yml
        Template-Provider-Unicode-Japanese.yml
        utf_16_le_bom.yml toolbar.yml vanilla.yml)
  ),
);

sub usage {
  print <<HERE;
Usage: $0 -b /path/to/blead/checkout
Does a diff of various files from here to blead
-r reverses the diff
HERE
  exit(1);
}

my $bleadpath;
my $reverse = 0;
GetOptions(
  'b|blead=s' => \$bleadpath,
  'h|help' => \&usage,
  'r|reverse' => \$reverse,
);

usage() if not defined $bleadpath or not -d $bleadpath;

foreach my $local_file (keys %files) {
  my $blead_file = "$bleadpath/" . $files{$local_file};
  my @cmd = ($diff, '-u');
  if ($reverse) {
    push @cmd, $blead_file, $local_file;
  }
  else {
    push @cmd, $local_file, $blead_file;
  }
  my $result = `@cmd`;
  my $blead_prefix = quotemeta($reverse ? '---' : '+++');
  my $local_prefix = quotemeta($reverse ? '+++' : '---');
  $result =~ s/^($blead_prefix\s*)(\S+)/$1.fix_blead_path($2,$bleadpath)/gme;
  $result =~ s/^($local_prefix\s*)(\S+)/$1.$files{$2}/gme;
  print $result;
}

sub fix_blead_path {
  my $path = shift;
  my $bleadpath = shift;
  $path =~ s/^\Q$bleadpath\E//;
  $path =~ s/^\/+//;
  return $path;
}



lib/Parse/CPAN/Meta.pm
use 5.008001;
use strict;
package Parse::CPAN::Meta;
# ABSTRACT: Parse META.yml and META.json CPAN metadata files

our $VERSION = '1.4423';

use Exporter;
use Carp 'croak';

our @ISA = qw/Exporter/;
our @EXPORT_OK = qw/Load LoadFile/;

sub load_file {
  my ($class, $filename) = @_;

  my $meta = _slurp($filename);

  if ($filename =~ /\.ya?ml$/) {
    return $class->load_yaml_string($meta);
  }
  elsif ($filename =~ /\.json$/) {
    return $class->load_json_string($meta);
  }
  else {
    $class->load_string($meta); # try to detect yaml/json
  }
}

sub load_string {
  my ($class, $string) = @_;
  if ( $string =~ /^---/ ) { # looks like YAML
    return $class->load_yaml_string($string);
  }
  elsif ( $string =~ /^\s*\{/ ) { # looks like JSON
    return $class->load_json_string($string);
  }
  else { # maybe doc-marker-free YAML
    return $class->load_yaml_string($string);
  }
}

sub load_yaml_string {
  my ($class, $string) = @_;
  my $backend = $class->yaml_backend();
  my $data = eval { no strict 'refs'; &{"$backend\::Load"}($string) };
  croak $@ if $@;
  return $data || {}; # in case document was valid but empty
}

sub load_json_string {
  my ($class, $string) = @_;
  require Encode;
  # load_json_string takes characters, decode_json expects bytes
  my $encoded = Encode::encode('UTF-8', $string, Encode::PERLQQ());
  my $data = eval { $class->json_decoder()->can('decode_json')->($encoded) };
  croak $@ if $@;
  return $data || {};
}

sub yaml_backend {
  if (! defined $ENV{PERL_YAML_BACKEND} ) {
    _can_load( 'CPAN::Meta::YAML', 0.011 )
      or croak "CPAN::Meta::YAML 0.011 is not available\n";
    return "CPAN::Meta::YAML";
  }
  else {
    my $backend = $ENV{PERL_YAML_BACKEND};
    _can_load( $backend )
      or croak "Could not load PERL_YAML_BACKEND '$backend'\n";
    $backend->can("Load")
      or croak "PERL_YAML_BACKEND '$backend' does not implement Load()\n";
    return $backend;
  }
}

sub json_decoder {
  if (my $decoder = $ENV{CPAN_META_JSON_DECODER}) {
    _can_load( $decoder )
      or croak "Could not load CPAN_META_JSON_DECODER '$decoder'\n";
    $decoder->can('decode_json')
      or croak "No decode_json sub provided by CPAN_META_JSON_DECODER '$decoder'\n";
    return $decoder;
  }
  return $_[0]->json_backend;
}

sub json_backend {
  if (my $backend = $ENV{CPAN_META_JSON_BACKEND}) {
    _can_load( $backend )
      or croak "Could not load CPAN_META_JSON_BACKEND '$backend'\n";
    $backend->can('new')
      or croak "No constructor provided by CPAN_META_JSON_BACKEND '$backend'\n";
    return $backend;
  }
  if (! $ENV{PERL_JSON_BACKEND} or $ENV{PERL_JSON_BACKEND} eq 'JSON::PP') {
    _can_load( 'JSON::PP' => 2.27300 )
      or croak "JSON::PP 2.27300 is not available\n";
    return 'JSON::PP';
  }
  else {
    _can_load( 'JSON' => 2.5 )
      or croak  "JSON 2.5 is required for " .
                "\$ENV{PERL_JSON_BACKEND} = '$ENV{PERL_JSON_BACKEND}'\n";
    return "JSON";
  }
}

sub _slurp {
  require Encode;
  open my $fh, "<:raw", "$_[0]" ## no critic
    or die "can't open $_[0] for reading: $!";
  my $content = do { local $/; <$fh> };
  $content = Encode::decode('UTF-8', $content, Encode::PERLQQ());
  return $content;
}
  
sub _can_load {
  my ($module, $version) = @_;
  (my $file = $module) =~ s{::}{/}g;
  $file .= ".pm";
  return 1 if $INC{$file};
  return 0 if exists $INC{$file}; # prior load failed
  eval { require $file; 1 }
    or return 0;
  if ( defined $version ) {
    eval { $module->VERSION($version); 1 }
      or return 0;
  }
  return 1;
}

# Kept for backwards compatibility only
# Create an object from a file
sub LoadFile ($) { ## no critic
  return Load(_slurp(shift));
}

# Parse a document from a string.
sub Load ($) { ## no critic
  require CPAN::Meta::YAML;
  my $object = eval { CPAN::Meta::YAML::Load(shift) };
  croak $@ if $@;
  return $object;
}

1;

__END__

=pod

=for Pod::Coverage

=head1 SYNOPSIS

    #############################################
    # In your file
    
    ---
    name: My-Distribution
    version: 1.23
    resources:
      homepage: "http://example.com/dist/My-Distribution"
    
    
    #############################################
    # In your program
    
    use Parse::CPAN::Meta;
    
    my $distmeta = Parse::CPAN::Meta->load_file('META.yml');
    
    # Reading properties
    my $name     = $distmeta->{name};
    my $version  = $distmeta->{version};
    my $homepage = $distmeta->{resources}{homepage};

=head1 DESCRIPTION

B<Parse::CPAN::Meta> is a parser for F<META.json> and F<META.yml> files, using
L<JSON::PP> and/or L<CPAN::Meta::YAML>.

B<Parse::CPAN::Meta> provides three methods: C<load_file>, C<load_json_string>,
and C<load_yaml_string>.  These will read and deserialize CPAN metafiles, and
are described below in detail.

B<Parse::CPAN::Meta> provides a legacy API of only two functions,
based on the YAML functions of the same name. Wherever possible,
identical calling semantics are used.  These may only be used with YAML sources.

All error reporting is done with exceptions (die'ing).

Note that META files are expected to be in UTF-8 encoding, only.  When
converted string data, it must first be decoded from UTF-8.

=head1 METHODS

=head2 load_file

  my $metadata_structure = Parse::CPAN::Meta->load_file('META.json');

  my $metadata_structure = Parse::CPAN::Meta->load_file('META.yml');

This method will read the named file and deserialize it to a data structure,
determining whether it should be JSON or YAML based on the filename.
The file will be read using the ":utf8" IO layer.

=head2 load_yaml_string

  my $metadata_structure = Parse::CPAN::Meta->load_yaml_string($yaml_string);

This method deserializes the given string of YAML and returns the first
document in it.  (CPAN metadata files should always have only one document.)
If the source was UTF-8 encoded, the string must be decoded before calling
C<load_yaml_string>.

=head2 load_json_string

  my $metadata_structure = Parse::CPAN::Meta->load_json_string($json_string);

This method deserializes the given string of JSON and the result.  
If the source was UTF-8 encoded, the string must be decoded before calling
C<load_json_string>.

=head2 load_string

  my $metadata_structure = Parse::CPAN::Meta->load_string($some_string);

If you don't know whether a string contains YAML or JSON data, this method
will use some heuristics and guess.  If it can't tell, it assumes YAML.

=head2 yaml_backend

  my $backend = Parse::CPAN::Meta->yaml_backend;

Returns the module name of the YAML serializer. See L</ENVIRONMENT>
for details.

=head2 json_backend

  my $backend = Parse::CPAN::Meta->json_backend;

Returns the module name of the JSON serializer.  If C<CPAN_META_JSON_BACKEND>
is set, this will be whatever that's set to.  If not, this will either
be L<JSON::PP> or L<JSON>.  If C<PERL_JSON_BACKEND> is set,
this will return L<JSON> as further delegation is handled by
the L<JSON> module.  See L</ENVIRONMENT> for details.

=head2 json_decoder

  my $decoder = Parse::CPAN::Meta->json_decoder;

Returns the module name of the JSON decoder.  Unlike L</json_backend>, this
is not necessarily a full L<JSON>-style module, but only something that will
provide a C<decode_json> subroutine.  If C<CPAN_META_JSON_DECODER> is set,
this will be whatever that's set to.  If not, this will be whatever has
been selected as L</json_backend>.  See L</ENVIRONMENT> for more notes.

=head1 FUNCTIONS

For maintenance clarity, no functions are exported by default.  These functions
are available for backwards compatibility only and are best avoided in favor of
C<load_file>.

=head2 Load

  my @yaml = Parse::CPAN::Meta::Load( $string );

Parses a string containing a valid YAML stream into a list of Perl data
structures.

=head2 LoadFile

  my @yaml = Parse::CPAN::Meta::LoadFile( 'META.yml' );

Reads the YAML stream from a file instead of a string.

=head1 ENVIRONMENT

=head2 CPAN_META_JSON_DECODER

By default, L<JSON::PP> will be used for deserializing JSON data.  If the
C<CPAN_META_JSON_DECODER> environment variable exists, this is expected to
be the name of a loadable module that provides a C<decode_json> subroutine,
which will then be used for deserialization.  Relying only on the existence
of said subroutine allows for maximum compatibility, since this API is
provided by all of L<JSON::PP>, L<JSON::XS>, L<Cpanel::JSON::XS>,
L<JSON::MaybeXS>, L<JSON::Tiny>, and L<Mojo::JSON>.

=head2 CPAN_META_JSON_BACKEND

By default, L<JSON::PP> will be used for deserializing JSON data.  If the
C<CPAN_META_JSON_BACKEND> environment variable exists, this is expected to
be the name of a loadable module that provides the L<JSON> API, since
downstream code expects to be able to call C<new> on this class.  As such,
while L<JSON::PP>, L<JSON::XS>, L<Cpanel::JSON::XS> and L<JSON::MaybeXS> will
work for this, to use L<Mojo::JSON> or L<JSON::Tiny> for decoding requires
setting L</CPAN_META_JSON_DECODER>.

=head2 PERL_JSON_BACKEND

If the C<CPAN_META_JSON_BACKEND> environment variable does not exist, and if
C<PERL_JSON_BACKEND> environment variable exists, is true and is not
"JSON::PP", then the L<JSON> module (version 2.5 or greater) will be loaded and
used to interpret C<PERL_JSON_BACKEND>.  If L<JSON> is not installed or is too
old, an exception will be thrown.  Note that at the time of writing, the only
useful values are 1, which will tell L<JSON> to guess, or L<JSON::XS> - if
you want to use a newer JSON module, see L</CPAN_META_JSON_BACKEND>.

=head2 PERL_YAML_BACKEND

By default, L<CPAN::Meta::YAML> will be used for deserializing YAML data. If
the C<PERL_YAML_BACKEND> environment variable is defined, then it is interpreted
as a module to use for deserialization.  The given module must be installed,
must load correctly and must implement the C<Load()> function or an exception
will be thrown.

=cut


Makefile.PL
use strict;
use warnings;

use 5.008001;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Parse META.yml and META.json CPAN metadata files",
  "AUTHOR" => "Adam Kennedy <adamk\@cpan.org>, David Golden <dagolden\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Parse-CPAN-Meta",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.008001",
  "NAME" => "Parse::CPAN::Meta",
  "PREREQ_PM" => {
    "CPAN::Meta::YAML" => "0.011",
    "Carp" => 0,
    "Encode" => 0,
    "Exporter" => 0,
    "File::Spec" => "0.80",
    "JSON::PP" => "2.27300",
    "strict" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => "0.80",
    "File::Spec::Functions" => 0,
    "JSON::PP" => "2.27300",
    "Test::More" => "0.47",
    "lib" => 0,
    "utf8" => 0,
    "vars" => 0,
    "warnings" => 0
  },
  "VERSION" => "1.4423",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "CPAN::Meta::YAML" => "0.011",
  "Carp" => 0,
  "Encode" => 0,
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => "0.80",
  "File::Spec::Functions" => 0,
  "JSON::PP" => "2.27300",
  "Test::More" => "0.47",
  "lib" => 0,
  "strict" => 0,
  "utf8" => 0,
  "vars" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

# Added by Dist::Zilla::Plugin::MakeMaker::Highlander
if ( $] < 5.012
  && ! $ENV{PERL_NO_HIGHLANDER}
  && ! ( $ENV{PERL_MM_OPT} && $ENV{PERL_MM_OPT} =~ /(?:INSTALL_BASE|PREFIX)/ )
  && ! grep { /INSTALL_BASE/ || /PREFIX/ } @ARGV
) {
    $WriteMakefileArgs{UNINST} = 1;
}

$WriteMakefileArgs{INSTALLDIRS} = 'perl'
    if $] >= 5.010001 && $] <= 5.011000;

WriteMakefile(%WriteMakefileArgs);
